omnimove_external_control/src/omnimove_external_control.cpp

```diff
@@ -0,0 +1,357 @@
+  if (
+    info.hardware_parameters.find("velocity_command_timeout_ms") != info.hardware_parameters.end())
+  {
+    vel_cmd_timeout_ms_ = std::stoi(info_.hardware_parameters["velocity_command_timeout_ms"]);
+  }
+  else
+  {
+    RCLCPP_INFO(
+      rclcpp::get_logger("OmnimoveExternalControl"), "velocity timeout not set. Using 500 ms");
+
+    vel_cmd_timeout_ms_ = 500;
+  }
+  RCLCPP_INFO(
+    rclcpp::get_logger("OmnimoveExternalControl"), "port of external control server:%d",
+    external_control_port_);
+  return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
+}
+
+hardware_interface::CallbackReturn OmnimoveExternalControl::on_configure(
+  const rclcpp_lifecycle::State & previous_state)
+{
+  return SystemInterface::on_configure(previous_state);
+}
+
+hardware_interface::CallbackReturn OmnimoveExternalControl::on_cleanup(
+  const rclcpp_lifecycle::State & previous_state)
+{
+  return SystemInterface::on_cleanup(previous_state);
+}
+
+hardware_interface::CallbackReturn OmnimoveExternalControl::on_shutdown(
+  const rclcpp_lifecycle::State & previous_state)
+{
+  // TODO: need to stop listening here
+  // client_socket_->shutdown ();
+  client_socket_->shutdown(client_socket_->shutdown_both);
+  acceptor_->close();
+  return SystemInterface::on_shutdown(previous_state);
+}
+
+CallbackReturn OmnimoveExternalControl::on_activate(const rclcpp_lifecycle::State & previous_state)
+{
+  //        client_socket_ = std::make_shared<tcp::socket>(io_service_);
+  //      client_socket_->connect(tcp::endpoint(address::from_string(external_control_host_),
+  //      external_control_port_));
+  try
+  {
+    acceptor_.reset(
+      new tcp::acceptor(io_context_, tcp::endpoint(tcp::v4(), external_control_port_)));
+    // tcp::acceptor *accepter = new tcp::acceptor(io_context, tcp::endpoint(tcp::v4(),
+    // external_control_port_));
+  }
+  catch (std::exception & e)
+  {
+    RCLCPP_ERROR(
+      rclcpp::get_logger("OmnimoveExternalControl"), "Failed to start acceptor %s %d", e.what(),
+      external_control_port_);
+    return CallbackReturn::ERROR;
+  }
+
+  RCLCPP_INFO(
+    rclcpp::get_logger("OmnimoveExternalControl"), "Waiting for External Control Connection");
+  // can i wait for a maximum amount of time
+  try
+  {
+    client_socket_ = std::make_unique<tcp::socket>(io_context_);
+  }
+  catch (std::exception & e)
+  {
+    RCLCPP_ERROR(
+      rclcpp::get_logger("OmnimoveExternalControl"), "Failed to initialise socket %s", e.what());
+    return CallbackReturn::ERROR;
+  }
+  try
+  {
+    acceptor_->accept(*(client_socket_.get()));
+  }
+  catch (std::exception & e)
+  {
+    RCLCPP_ERROR(
+      rclcpp::get_logger("OmnimoveExternalControl"), "Failed to accept connection %s", e.what());
+    return CallbackReturn::ERROR;
+  }
+  RCLCPP_ERROR(rclcpp::get_logger("OmnimoveExternalControl"), "Accepted connection");
+
+  return SystemInterface::on_activate(previous_state);
```

Comment:
```suggestion
  return CallbackReturn::SUCCESS;
```

Reply:
same as the other two comments. What do you think?

Reply:
I do not think the base class will be ever extended with checks, as the condition for the transitions (whether the requested transition is valid) is checked before the callback is even called.
However this does not really cause an overhead, so it can remain as it is, if you prefer it this way :)